Update Mixtrack (Pro) 3 mappings for Mixxx 2.1

radusuciu opened the pull request:
This is a WIP, not quite ready for review since I haven't had time to actually test it. Opening pull request a bit early just to move discussion from #1014. First time contributor so please let me know if I'm going against the grain!

To-do:
- [x] Load sample if none is currently loaded
- [x] Make use of `focused_effect`
- [x] Map sampler eject
- [x] Look at bpm tap functionality, which I either broke, or is broken

radusuciu committed:
Load sampler if none is currently loaded.

- Pressing sampler button if no sample is currently loaded will load
sample from library
- Shift + sampler button if sample is loaded will eject appropriate
sample

Fix checking if track is loaded.

Make sure sample is stopped before ejecting

radusuciu committed:
Small refactor

radusuciu committed:
Add configurable beatjump size.

radusuciu committed:
Map loop move

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -858,6 +859,7 @@ NumarkMixtrack3.deck = function(decknum) {
     // NMTP3 is a "Model A" controller for scratching, it centers on 0.
     // See http://www.mixxx.org/wiki/doku.php/midi_scripting#scratching
     // and see "Jogger" object constructor
+    this.beatJumpSize = 1;
```
In the future, this will be available as a ControlObject that will also be shown on skins so controller scripts won't have to keep track of this. I might implement that for Mixxx 2.2. There is a [Launchpad ticket](https://bugs.launchpad.net/mixxx/+bug/1429331) tracking it.

radusuciu replied:
That would be awesome. I will not be committing code that will modify this value until the CO will be there as this controller has no way of displaying it.

Be-ing replied:
You may find a beatjump size of 4 beats to be more helpful.

djsteph replied:
FYI, I just tried to install Mixxx 2.1 on Windows 10 and the installer failed. Also, I got 3 different warnings asking to confirm that I wanted to install the package; the first one saying that Windows did not recommend installing it, likely because package was not signed.

radusuciu replied:
In Windows 10 you have to click "More Info" and then tell it you really mean to install it.

Be-ing replied:
Please [report a bug](https://bugs.launchpad.net/mixxx/+filebug) for that.

radusuciu committed:
Avoid code duplication for 4 deck support

radusuciu committed:
Use new track_loaded control.

radusuciu commented:
Will the new effects interface be used across all skins? There is currently a lot of code duplication as a result of having to support skins that have only one effect per rack. Documented here:

```javascript
// Effects mode
// 1 for multi-effect mode (3 effects controlled in each EffectUnit, best used with Deere Skin)
// 2 for single effect mode (1 effect controlled in each EffectUnit, best used with skins other than Deere)
var FXmode = 1; // multi-effect mode by default
```

Be-ing commented:
> Will the new effects interface be used across all skins?

Yes.

radusuciu commented:
Awesome, thank you. 👍

radusuciu committed:
Remove checking for effect rack configuration

radusuciu committed:
Simplify LED init

radusuciu committed:
Remove softTakeover workaround

radusuciu committed:
Bump mininum mixxx version to 2.1

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1085,14 +1052,28 @@ NumarkMixtrack3.init = function(id, debug) {
     }
 
-    for (i = engine.getValue("[Master]", "num_decks"); i >= 1; i--) {
+    for (var i = engine.getValue("[Master]", "num_decks"); i >= 1; i--) {
```
FYI, in the old version of JS that Mixxx still uses (and in modern JS using `var`), variables are scoped to functions, not blocks as you might expect. You don't need to redeclare `i` for each loop. Just something to be aware of if you encounter puzzling bugs with one loop messing up the behavior of the next.

radusuciu replied:
I'm aware that JS lacks block scoping when using `var`, and that variable declarations get hoisted, but I think pretty much everyone except Crockford will advocate for including var in for loops. 

As an aside, the mapping in its current state is not functional since I committed stuff without having the controller on hand. I will go back and test things commit by commit though!

Be-ing replied:
Okay, just pointing it out because many people who contribute mapping scripts aren't that familiar with JS.

radusuciu committed:
Fix LED init

radusuciu committed:
Fix TrackIsLoaded call

radusuciu committed:
Refactor deckFromGroup

radusuciu committed:
Finish removing shift lock option

radusuciu committed:
Remove fastseekenabled option

Saw this being discussed in previous PR. This makes sense as a default,
especially since it requires a modifer and shift lock is no longer in
the code.

radusuciu commented:
@djsteph Just came across this: https://docs.google.com/spreadsheets/d/1YbLS7hBprP2uvEKizO-b0WiaRmpvml0r5NfBp4GmBVc/edit#gid=0

I'm guessing it was most useful during development but I think I'd like to use it as well. It will make a fantastic quality check for the mapping.

radusuciu committed:
Refactor wheel turn handler

radusuciu committed:
FIx mistake in deckFromGroup

deckGroup is necessary because it is used when decks are switched

radusuciu committed:
Refactor track end warning handling

djsteph commented:
You are welcome to use it...I think anyone can edit the file

Stéphane Morin
Maison: 450 510-0622
Cellulaire: 514 243-5876

On Fri, Feb 10, 2017 at 4:23 PM, Radu Suciu <notifications@github.com>
wrote:

> @djsteph <https://github.com/djsteph> Just came across this:
> https://docs.google.com/spreadsheets/d/1YbLS7hBprP2uvEKizO-
> b0WiaRmpvml0r5NfBp4GmBVc/edit#gid=0
>
> I'm guessing it was most useful during development but I think I'd like to
> use it as well. It will make a fantastic quality check for the mapping.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1180#issuecomment-279069548>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQDKyQgEOAPE4mdCerrXSVUkS_pG51EBks5rbNVcgaJpZM4L81Rl>
> .
>

radusuciu commented:
Trying to debug things and getting a billion messages of this nature:

```
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 184549449 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 268435486 
Debug [Controller]: Killing timer: 184549449 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: Killing timer: 268435486 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x00" 

Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 1493172255 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 1509949470 
Debug [Controller]: Killing timer: 1493172255 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: Killing timer: 1509949470 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x00" 

Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 805306398 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 922746911 

Debug [Controller]: Killing timer: 805306398 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: Killing timer: 922746911 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 1811939393 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 1912602654 

Debug [Controller]: Killing timer: 1811939393 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: Killing timer: 1912602654 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x92 (ch 3, opcode 0x9), ctrl 0x13, val 0x00" 
Debug [Controller]: "MixTrack Pro 3: outgoing: status 0x91 (ch 2, opcode 0x9), ctrl 0x13, val 0x20" 
Debug [Controller]: Starting one-shot timer: 1660944414 
```

Is this normal?

Be-ing commented:
Yes, that is normal. I added outgoing MIDI debugging messages in #1004. Is that an issue? Would you like an option to only see incoming MIDI?

radusuciu committed:
Remove accidental semicolon

radusuciu commented:
Well, it kind of makes it difficult to debug by logging things with `print` statements since anything you log gets drowned out very quickly by messages from permanent timers.

Perhaps I'm doing debugging wrong? These are the options I am using:

```
--developer --debugLevel 2 --controllerDebug --resourcePath "%USERPROFILE%/Github/mixxx/res"
```

Be-ing commented:
``--debugLevel`` was recently renamed to ``--logLevel`` (#1159), so that does not do anything currently, but ``--developer`` automatically sets logLevel to debug unless otherwise specified. Try running with ``--logLevel debug`` without ``--controllerDebug``. That might hide the MIDI debugging messages but still let scripts print to the console output.

radusuciu committed:
Compress revision history

radusuciu committed:
Refactor brake and spinback effect handling

radusuciu committed:
Small refactor for style

radusuciu committed:
Refactor connectDeckControls method

radusuciu committed:
Get track duration directly from engine

I believe this work-around is no longer necessary after this bug was
fixed: https://bugs.launchpad.net/mixxx/+bug/1497183 in
https://github.com/mixxxdj/mixxx/pull/970

radusuciu committed:
refactor BeatKnob method

radusuciu committed:
Remove range checking before setting parameter

Not needed with setParameter

radusuciu committed:
Remove DarkMetal skin workaround

As per https://github.com/mixxxdj/mixxx/pull/1179, workaround should not
be necessary any more.

djsteph commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1244,29 +1240,15 @@ NumarkMixtrack3.BrowseButton = function(channel, control, value, status, group)
-        contract = 0;
-    }
-
     if (shifted && value === ON) {
```
Just curious, you are not keeping the maximize library on the browse button push? In this code sniplet, I don't see any value set for LibraryGroup and Library command nor do I see the actual change in the code to remove the variable

radusuciu replied:
That was an oversight - thanks for noticing. Will fix.

radusuciu committed:
Fix browse button handling.

radusuciu committed:
Fix small errors

radusuciu committed:
Minor refactor. Syntax and style.

radusuciu committed:
Loop deck and sampler init.

radusuciu committed:
Minor refactor

radusuciu committed:
Move strip touch handling to one place

Was the only method on the deck protoype. In the future, perhaps the
reverse move would be smart as it would eliminate a lot of
```deckFromGroup``` calls, but it may require a reworking of the 4-deck
functionality.

radusuciu committed:
Use toggleValue instead of manually toggling

radusuciu committed:
Fix mistakes.

radusuciu commented:
Having some trouble with doing soft takeover when using modifier keys + EQ/Filter knobs. Desired functionality (abridged from wiki):

> 9. Filter
> Adjusts the amount of the filter effect. Turning the knob left controls the low pass filter; turning it right controls the high pass filter.
> Shift + Filter: Adjust FX 1, parameter 4 of the Effect Unit assigned to the deck

Is there a way to do soft takeover between Filter and Shift + Filter that is built in, or do I need to add this back:

```javascript
function parameterSoftTakeOver(group, control, value) {        
    var threshold = 0.07; //on the CMD Studio 4a this threshold got the right balance         
    //between smooth takeover and keeping up with quick turns, but you can adjust the value to suit your needs        
    var currentKnobVal = value / 127;     
    var currentParamVal = engine.getParameter(group, control);        
    var spread = Math.abs(currentParamVal - currentKnobVal);      
      
    if (spread < threshold) {     
        engine.setParameter(group, control, currentKnobVal); //set the new value      
    } else {      
        return; //do nothing until we get close       
    }     
}
```

Using ```engine.softTakeover``` in the init works fine when manipulating the UI without the controller.. I've been experimenting with ```engine.softTakeoverIgnoreNextValue``` but haven't hit it yet.

Thanks!

radusuciu committed:
Fix bug when sample button is held

Be-ing commented:
Maybe you are calling `engine.softTakeoverIgnoreNextValue` on the wrong ControlObject? I think it should be called for the CO that was being manipulated and is being switched away from, not the CO that is being switched to.

radusuciu commented:
Tried every permutation of:

```javascript
    // default behavior is to control filter
    if (!deck.shiftKey && !deck.PADMode && !deck.TapDown) {
        engine.setParameter("[QuickEffectRack1_[Channel" + decknum + "]]", "super1", knobValue);
        engine.softTakeoverIgnoreNextValue("[EffectRack1_EffectUnit" + decknum + "_Effect1]", "parameter4");
    }

    // modified behaviour controls effect parameters
    if (deck.shiftKey) {
        engine.setParameter("[EffectRack1_EffectUnit" + decknum + "_Effect1]", "parameter4", knobValue);
        engine.softTakeoverIgnoreNextValue("[QuickEffectRack1_[Channel" + decknum + "]]", "super1");
    }
```
And it doesn't work, while ``parameterSoftTakeOver``` the functions works.

I need to see if I can understand the C++ but I think what is needed in this case is more likely to be expressed as ```softTakeoverUntilClose```.

Be-ing commented:
Hm, maybe the issue is that you're calling `engine.softTakeoverIgnoreNextValue` upon each MIDI input message received for the knob. Try moving that to the function that handles input for the shift button.

> I need to see if I can understand the C++

I've glanced at it... it's really complicated.

radusuciu commented:
Still didn't work when moved one `softTakeoverIgnoreNextValue` or both calls into the `ShiftButton` method. Maybe I'm still doing something wrong but I think that restoring `parameterSoftTakeover` is the most pragmatic solution.

radusuciu committed:
Revert parameterSoftTakeOver refactor

radusuciu committed:
Simplify calculation of effect and eq param numbers

radusuciu committed:
Make use of focused effect in changing fx params

Still haven't coded in the part where we actually set the focused
effect, but I'm playing around with what's been suggested in Steph's
last pull request.

This will allow for control of individual effect parameters without
having to use 3 different modifier keys. Now, just Shift + EQ/Filter
knob on focused effect. This introduces the extra step of focusing an
effect, but if this is made easy enough, I think it will be more usable.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1986,45 +1986,35 @@ NumarkMixtrack3.bpmTap = function(channel, control, value, status, group) {
 NumarkMixtrack3.EQKnob = function(channel, control, value, status, group) {
     var deck = NumarkMixtrack3.deckFromGroup(group);
     var decknum = deck.decknum;
+    var focusedEffect = engine.getValue("[EffectRack1_EffectUnit" + decknum + "]", "focused_effect");
```
Make sure you set show_focus to 1 for every EffectUnit or the focus won't be shown on screen.

radusuciu committed:
Simplify library maximize toggle

radusuciu committed:
Fix bug in callback conditional

radusuciu committed:
Remove unnecessary properties and variables

radusuciu commented:
I think it'd be nice if Tap was re-mapped to toggle focus between controls. Would lose bpm tap but I don't think anyone really uses that..

Too confusing?

Be-ing commented:
If Mixxx's BPM detection was awful, I'd be more concerned about having a tap BPM button. Anyway, that's more a track preparation task. While it can be helpful to have those kinds of functions accessible on a controller (I occasionally use a beatgrid realignment button during a set), IMO mappings should be more focused on what it is important to have access to while mixing.

I have been thinking about reworking the UX for the EffectUnit object in Components in a similar way to use the 4th effect button for switching effect focus instead of having to reach all the way down to the shift button (which is typically not next to the effects section on most controllers). You could make the LED of the TAP button light up when an effect is focused, then pressing it could unfocus.

Figuring out the optimal workflow for handling focusing and Instant FX will probably take a while of trying different ways of mapping it. It doesn't have to be perfect immediately. So pick some way to do it, implement it, and play with it.

radusuciu commented:
Great, thanks!

Be-ing commented:
Just as a heads up...
https://xkcd.com/1172/
:P

radusuciu committed:
Fix tap button calls to engine

If I understand correctly, `bpm.tapButton` isn't necessary when using
the `bpm_tap` control.

radusuciu committed:
Use tap button to focus next effect

radusuciu committed:
Small refactor

radusuciu committed:
Use deck object directly in deck init

radusuciu committed:
Set meta for focused effect

radusuciu committed:
Allow simultaneous use of instant and focused fx

Also adjust focus toggling so that adding an instant effect does not
also change focus.

radusuciu committed:
Switch slip_enabled and quantize

Shifted to match wiki, comments and I think it also is better to have
slip more accessible than quantize.

radusuciu committed:
Refactor FXButton and move around modifiers

Pad + Fx = Instant
Shift + Fx = Select Effect
Tap + Fx = Focus effect

radusuciu committed:
Move sampler volume to PAD + Browse

radusuciu committed:
Instant FX to Pad, Focus to Tap

radusuciu committed:
Switch loop_move and beatjump

Now loop_move with shift + beats knob, beatjump with tap + beats knob

radusuciu committed:
Adjust loop_move size and beatjump size

Using existing [BeatJump] and [LoopMove] CO's to reflect these changes
in the UI.

radusuciu committed:
Fixes for beat jump/loop move + small mistakes

radusuciu committed:
Fix mistake in jump size determination

radusuciu committed:
Use beatjump_X control to reset beatjump value

Same for loop_move

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1814,13 +1819,17 @@ NumarkMixtrack3.FXButton = function(channel, control, value, status, group) {
     } else if (deck.TapDown && value === DOWN) {
-        deck.focusEffect(effectNum);
+        if (deck.focusedEffect() === effectNum) {
+            deck.focusedEffect(0);
```
This function name is confusing. Should it be `focusEffect` (not focus_ed_)?

radusuciu replied:
I thought the same as well, I chose `focusedEffect` because when calling it without any arguments it returns the focusedEffect. It's awkward though and I'll change it next time I have time to look at the code.

Be-ing replied:
Cool, I suggest splitting those into two functions.

radusuciu committed:
Focus appropriate sampler bank when adjusting gain

radusuciu committed:
Eject sample with Shift + Double Press

radusuciu committed:
Fix track eject

Without setting to false, the button skin remains in the "pressed"
state.

radusuciu committed:
Revert "Eject sample with Shift + Double Press"

This reverts commit 6883720d434a98b1882f17306cad207f2802ae14.

radusuciu committed:
Eject sample with Tap modifier

`LongShortDoubleBtn` did not behave as expected and Shift + Double press
wasn't the best mapping anyway.

radusuciu committed:
Split up `focusedEffect` method

radusuciu commented:
Seems that tapping the bpm button only changes the file_bpm control, not the actual bpm of the playing song.

radusuciu committed:
Smoother experience loading and ejecting

Be-ing commented:
Regarding BPM tapping: https://bugs.launchpad.net/mixxx/+bug/1441588

radusuciu committed:
Reset bpm tap button after engaging

radusuciu committed:
Map slip to shift + wheel, quantize to shift + pfl

radusuciu commented:
@Be-ing, is migrating mappings to your components lib a goal or is it kind of an "if it ain't broke don't fix it" situation?

Be-ing commented:
If it ain't broke, you don't have to fix it. You're welcome to fix it if you want to though.

Be-ing commented:
Do you intend on changing this further or should I review it now? Please note changes in functionality in a new section on [the wiki page](http://mixxx.org/wiki/doku.php/numark_mixtrack_pro_3).

radusuciu commented:
There's one more tweak (bugfix I think) for the wheeltouch handling that I want to take care of, but that will be a small and isolated set of changes.

Will add changes in a new section as requested.

radusuciu commented:
So, I can't make a new account on the wiki

Be-ing commented:
:/ Is there an error message it gives you? Do you never get a confirmation email? Is your wiki user name at least 3 characters?

radusuciu commented:
On Chrome, Doku kept insisting that I am a robot. For some reason I didn't encounter this issue when using Edge.

http://mixxx.org/wiki/doku.php/numark_mixtrack_pro_3#proposed_changes

radusuciu commented:
Btw, played for ~6 hours straight this weekend with no issues.

Be-ing commented:
Thanks for updating the wiki. I'll take a closer look soon.

radusuciu commented:
Kinda tempted to remove the auto-cut feature, is that a bad idea?

Be-ing commented:
Only remove it if there is a good reason to, like if you want to replace it with a different functionality. If it's not getting in the way, leave it.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2067,157 +1754,109 @@ NumarkMixtrack3.OnPADLoopButtonChange = function(value, group, control) {
+    } else if (deck.shiftKey && value === DOWN) {
+        // load next effect, make sure it is inactive, and make sure the unit is enabled
+        engine.setValue(effectGroup, "next_effect", true);
+        engine.setValue(effectGroup, "enabled", false);
```
I don't think this should assume the user wants to disable the effect. Loading different effects and keeping the slot enabled can be helpful for picking out which effect to use.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1819,119 +1547,95 @@ NumarkMixtrack3.WheelMove = function(channel, control, value, status, group) {
+
+            sampler.LEDs["PADsampler" + padIndex].flashOn(300, PADcolors.purple, 300);
         } else {
             engine.setValue(group, "stop", 1);
```
I think it's better to map samplers so pressing the button for a playing sampler goes back to the cue point and shift + a playing sampler stops it. The way it is mapped here, the sample would have to be stopped first before triggering it again. Can you set these pads to different colors? If so, use a different color for a playing sampler and a loaded but stopped sampler.

radusuciu replied:
The pads do turn different colors but I think they are limited to 3 which are used for manual, auto, and sampler modes so that might be confusing. I agree with your suggested change, but perhaps I can have the pad flash slowly while playing.

Be-ing replied:
> perhaps I can have the pad flash slowly while playing

Worth a try.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1819,119 +1547,95 @@ NumarkMixtrack3.WheelMove = function(channel, control, value, status, group) {
-                //play sampler with Sync
+                // play sampler with Sync
                 engine.setValue(group, "cue_gotoandplay", 1);
                 engine.setValue(group, "beatsync", 1);
```
I'm not sure samplers should always be synced. That could make it awkward to use short samples that don't align so well with the beatgrid.

radusuciu replied:
It was mapped as Shift + Sampler -> Play without sync, this conflicts with suggestion to map Shift + Sampler -> Stop.

Will have to remap or change behavior somehow.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2067,157 +1754,109 @@ NumarkMixtrack3.OnPADLoopButtonChange = function(value, group, control) {
+        if (focusedEffect) {
+            var focusedGroup = "[EffectRack1_EffectUnit" + deck.decknum + "_Effect" + focusedEffect + "]";
+            engine.setValue(focusedGroup, "meta", value / 127);
+        }
```
It looks like it is possible to have Instant FX and a focused effect. Should that state be possible?

radusuciu replied:
It was on purpose, I didn't want to decide which functionality gets priority. Can rethink this.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2253,358 +1888,222 @@ NumarkMixtrack3.PFLButton = function(channel, control, value, status, group) {
+NumarkMixtrack3.OnBeatJumpX = function(value, group, control) {
+    if (!value) return;
+    beatJumpSize = parseInt(control.match(/\d/)[0]);        
+};
```
I realized I should implement this in #1187. I already did for loop sizes, but didn't think to do so for the beatjump_X_forward/backward controls. So this shouldn't be necessary after that is merged.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2253,358 +1888,222 @@ NumarkMixtrack3.PFLButton = function(channel, control, value, status, group) {
+    // default behavior is to control EQ
+    // when shifted, change parameters of focused effect
+    if (deck.shiftKey && focusedEffect) {
+        parameterSoftTakeOver(
```
I'm still confused why engine.softTakeover wasn't working for you...

radusuciu replied:
I can revisit this.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2253,358 +1888,222 @@ NumarkMixtrack3.PFLButton = function(channel, control, value, status, group) {
 
     if (value === DOWN) {
-        engine.setValue("[Channel" + decknum + "]", "reloop_exit", value ? 1 : 0);
+        engine.setValue(deck.group, "reloop_exit", value ? 1 : 0);
```
You won't necessarily have to change this because the old reloop_exit name will be aliased, but #1187 will change this confusing name to reloop_toggle.

Be-ing reviewed (commented):
Overall looks quite good. I think there is a bit of room for improvement with the samplers and effects.

radusuciu commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2067,157 +1754,109 @@ NumarkMixtrack3.OnPADLoopButtonChange = function(value, group, control) {
+        // load next effect, make sure it is inactive, and make sure the unit is enabled
+        engine.setValue(effectGroup, "next_effect", true);
+        engine.setValue(effectGroup, "enabled", false);
+        engine.setValue("[EffectRack1_EffectUnit" + decknum + "]", "group_[Channel" + decknum + "]_enable", true);
```
Currently switching decks also switches effect racks. This works with the Deere 2.1 when displaying 4 effect racks.

Personally I think 4 effect racks is overkill.. I think I'll try and have a toggle for the effect rack being applied to the current deck.. I'm worried that it may end up being confusing though.

Be-ing replied:
Regardless of whether you find a way to squeeze deck assignment buttons in, I don't think the mapping should force effect units to be tied to a deck. #1103 changed the behavior of the deck assignment switches so that each effect unit is enabled on its corresponding deck if no saved state is found in mixxx.cfg (which will happen on first start of Mixxx 2.1 even for users who are upgrading), so there is no need for mappings to forcibly enable deck assignments. This allows users to switch the deck assignments while mixing, even if they have to use their mouse to do so. Changes that users make to the default deck assignments persist across restarts of Mixxx.

radusuciu commented:
Thanks for review! Will take care of when I have time.

radusuciu committed:
Do not disable effect when switching effects

radusuciu committed:
Make use of script.toggleControl

radusuciu committed:
Go to cue if sample is playing, stop with shift

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1378,7 +1378,7 @@ NumarkMixtrack3.OnLoadSelectedTrack = function(value, group, control) {
  *                 the same tempo, and decks that also have quantize
  *                 enabled will always have their beats lined up.
- * If the Sync Loack was previously activated, it just desactivate it,
+ * If Sync Lock was previously activated, it just desactivate it,
```
Typo: "deactivate"

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1624,15 +1624,18 @@ NumarkMixtrack3.SamplerButton = function(channel, control, value, status, group)
             } else {
-                // play sampler with Sync
                 engine.setValue(group, "cue_gotoandplay", 1);
                 engine.setValue(group, "beatsync", 1);
```
You removed the comment about playing with sync. Did you mean to remove the syncing?

radusuciu replied:
No, I just thought the code is pretty clear in this case. Shift + Sampler will play the sample without sync so it's pretty easy to switch between the two.

Earlier I thought there'd be an issue with mapping Shift + Sampler to stop a sample, but there isn't one since stopping is only done on playing samples and playing without sync is only done on non-playing samples.

Be-ing reviewed (commented):
A few quick comments

radusuciu committed:
Give instantfx priority over focused effect

radusuciu committed:
Small fixes

Be-ing commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -1,6 +1,6 @@
+<MixxxControllerPreset mixxxVersion="2.1.0+" schemaVersion="1">
     <info>
         <name>Numark Mixtrack (Pro) 3</name>
         <author>Stéphane Morin</author>
```
Feel free to mention yourself here.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2253,358 +1883,222 @@ NumarkMixtrack3.PFLButton = function(channel, control, value, status, group) {
-        for (i = 1; i <= 4; i++) {
+    var deck = NumarkMixtrack3.deckFromGroup(group);
+    // beat knobs sends 1 or 127 as value. If value = 127, turn is counterclockwise
+    var increase = (value !== 127);
```
I think `var increase = (value === 1);` would be clearer.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2635,71 +2127,42 @@ NumarkMixtrack3.OnVolumeChange = function(value, group, control) {
-    var decknum = script.deckFromGroup(group);
-    var deck = NumarkMixtrack3.decks["D" + decknum];
+    if (deck.loaded && TrackEndWarning) {
+        var timeremaining = duration * (1 - value);
```
inconsistent capitalization of variable name

Be-ing reviewed (commented):
Wow that was a lot of code changed! Thanks for the cleanup. I found a handful of other places in the diff where the code could be cleaned up a little bit plus a few UX suggestions. Looks almost ready to merge.

radusuciu commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -366,23 +322,22 @@ LED.prototype.flashOn = function(num_ms_on, value, num_ms_off, flashCount,
-            });
+        // permanent timer
+        this.flashTimer = engine.beginTimer(num_ms_on + num_ms_off, function() {
+            myself.flashOnceOn(false);
```
is this:

```javascript
engine.beginTimer(num_ms_on + num_ms_off, 'this.flashOnceOn(false)';
```

preferred over:

```javascript
engine.beginTimer(num_ms_on + num_ms_off, this.flashOnceOn.bind(this, false);
```

I get that I don't need to bind for `this` but I also have an argument I'd like to curry. Most examples I've seen use the string method but I feel like the latter is better practice.

Be-ing replied:
IMO the ability to pass strings instead of functions is an ugly hack that shouldn't have been implemented in the first place. I think this form would be best:

```
engine.beginTimer(num_ms_on + num_ms_off, function () {
    this.flashOnceOn(false);
});
```

radusuciu commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -483,32 +439,25 @@ var SingleDoubleBtn = function(Callback, DoublePressTimeOut) {
-                }, true);
+    if (this.ButtonTimer === 0) { // first press
+        this.ButtonTimer = engine.beginTimer(this.DoublePressTimeOut, function() {
+            myself.ButtonDecide();
```
Similar to my question above.

```javascript
engine.beginTimer(this.DoublePressTimeOut, 'this.ButtonDecide()');
```

vs.

```javascript
engine.beginTimer(this.DoublePressTimeOut, this.ButtonDecide);
```

radusuciu committed:
Use this context directly. #1196

radusuciu committed:
Fix some capitalization issues

radusuciu committed:
Camelcase LEDs

radusuciu committed:
Finish camelcasing LED objects

radusuciu committed:
Remove effectUnit channel toggle

radusuciu committed:
Small refactor

radusuciu committed:
Remove superfluous comments

radusuciu commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1241,57 +1028,35 @@ NumarkMixtrack3.initDeck = function(group, remove) {
+
+        for (var led in offDeck.LEDs) {
+            if (led.hasOwnProperty('onOff')) {
+                led.onOff(OFF);
```
I'm not too familiar with this code but I agree. However, I think that can be part of a future refactor of the LED/Button objects. Either that or leave it and consider using the components lib you developed.

Be-ing replied:
Sure, not a big deal.

radusuciu commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -10,25 +10,12 @@ var TrackEndWarning = true;
 var smartPFL = true;
 
 // use beatlooproll instead of beatloop
 var beatlooprollActivate = false;
```
I don't think those behaviours are the same, this option makes it so that all the loops set with the auto loop buttons are rolling. Also, the Shift behaviour isn't to toggle loopRoll, but rather to allow for the toggling of fractional beat loops (mapping 4 buttons to 8 loop sizes).

I understand your suggestion but I'm not quite sure if I want all of my <= 1 beat loops to be rolling. Will consider it, especially if others want the shifted behaviour you describe.

For now, I will remove the option since I can't imagine anyone wanting all of their loops to be rolling, especially without a way to toggle. Will be revisiting this when #1187 is merged.

Be-ing replied:
What is the use of loops <= 1 beat that aren't rolling?

radusuciu replied:
Say instead of rolling into a drop on the same track, you bring in another track instead, and you want to leave the other track where it is instead of having it slip past when you stop the loop.

I suppose there are other ways around that though..

radusuciu committed:
Remove beatlooprollActivate option

radusuciu committed:
Camelcase padindex

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -614,12 +603,15 @@ LongShortDoubleBtn.prototype.ButtonDown = function(channel, control, value, stat
-        }, true);
+
+        this.ButtonLongPressTimer = engine.beginTimer(
+            this.LongPressThreshold, this.ButtonAssertLongPress, true
```
Hmm, I think this actually would have worked in Mixxx 2.0. What #1196 fixed was using `this` inside the callback function. For example: fd9815dd042e1a33b717d49db067271cef458c96

Be-ing replied:
Oh, the ButtonAssertLongPress function refers to `this`, so maybe it wouldn't have worked. Anyway, looks good now.

radusuciu replied:
Yeah, just figured there's little point in wrapping it with a function if no arguments need to be passed.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -155,25 +155,25 @@ var leds = {
     "hotCue3": 0x1d,
     "hotCue4": 0x1e,
-    "Cue": 0x03,
+    "cue": 0x03,
```
FYI, if you forget something in a commit and haven't added any new commits yet, `git commit --amend` is an easy way to overwrite the incomplete commit.

Be-ing commented:
Alright this looks good to me! 👍 Thank you! I'm glad one of the best selling controllers on the market right now will have a good mapping for the new effects UI for the 2.1 release. :)

After #1187 is merged you may want to reconsider how loops are mapped, but I've been careful to make sure that old mappings will still work with the new looping interface.

radusuciu commented:
Thanks! I want to do a final test before merge so as to not have to open another PR just for minor fixes.

I will follow up with a PR for #1187 and #940 in the near future.

Be-ing commented:
If you'd like to play with those before they're merged, you can [checkout PRs locally](https://gist.github.com/piscisaureus/3342247) and build them. I put some [tips](http://mixxx.org/wiki/doku.php/using_git#working_on_mappings_and_skins_separately_from_other_changes) on the wiki for using git worktree to play with experimental skins and mappings.

radusuciu committed:
Misc bug fixes for beatjump/loopmove

radusuciu committed:
Move brake and spinback

Conflict with InstantFX

radusuciu committed:
Adjust VUMeter multiplier

120 goes into red way too easily. 82 was juuuust right.

radusuciu commented:
Good to go from my end with these minor fixes. Effect units definitely need to be reworked though, will try and see if I can get a build going as you described.

Thanks for review and hand-holding!

Be-ing commented:
Could someone merge this?

I do not know if you are on the mailing list, but I just wrote a [proposal for registering MIDI inputs from JS](http://mixxx.org/wiki/doku.php/registering_midi_input_handlers_from_javascript). Feedback would be appreciated.

Be-ing commented:
@daschuer can you merge this?

daschuer commented:
Yo! Thank your for the work.

Be-ing commented:
@radusuciu, now that #1187 and #940 have been merged, could you update this for the new looping and beatjumping Controls? I suggest remapping the autoloop pad layer similar to the [Pioneer DDJ-SB2](https://mixxx.org/wiki/doku.php/pioneer_ddj-sb2#pads_-_auto_loop_mode). Users have been reporting that mapping pads that way is a little strange at first but great after getting used to it. Unlike the DDJ-SB2, because there are other parts of the controller that could be used for beatjumping, I think it would make more sense to use beatjump_1_backward/forward as the shift action for loop halve/double. You could use shift + pitch bend buttons to halve/double beatjump_size and shift + Beats encoder for beatjump_backward/forward.

Also, when I played with a DDJ-SB2, I found it was more helpful to have quick access to changing the focused effect than toggling the effect enable switches. I suggest switching these around in the Mixtrack Pro 3 mapping, so pressing the effect buttons changes the focused effect and Tap + effect button toggles the effect enable switches.

radusuciu commented:
Yep, I will update it to make use of the new controls and will consider your mapping suggestions. I've also done a bit of playing with your components lib so can maybe bring in `EffectUnit` to start.

Be-ing commented:
The EffectUnit object from the Components library is made for controllers with 4 effects knobs, so adapting it for the Mixtrack Pro 3 would probably be just as much (or maybe more) work than writing the code from scratch. However, using the Components and ComponentContainer objects to write an EffectUnit object specific to this controller would make sense if you'd like to rewrite the code you already have.

I don't recommend copying how the effects buttons work with the DDJ-SB2 in #1243. I used short versus long button presses to distinguish between switching focus and enabling effects, which is a bit clumsy because the script has to wait for the button to be released before switching the focus. There also isn't a way to show which effects are enabled on the controller. These don't need to be issues with the Mixtrack Pro 3 because the Tap button can be used to switch layers.
